cmake/clang_dev_tools.cmake

```diff
@@ -0,0 +1,53 @@
+# Run a set of LLVM/Clang tools on the code base to maintain high
+# code standards and catch potential security vulnerabilities
+#
+# Tools ran:
+# - clang-format
+# - clang-tidy
+# - include-what-you-use
+
+# Get c++ files across all targets
+file(
+  GLOB_RECURSE
+  ALL_SOURCE_FILES
+  libzeth/**.?pp libzeth/**.tcc
+  prover_server/**.?pp prover_server/**.tcc
+  mpc_tools/**.?pp mpc_tools/**.tcc
+)
+
+# Adding clang-format target if executable is found
+find_program(CLANG_FORMAT "clang-format")
+if(CLANG_FORMAT)
+  add_custom_target(
+    clang-format
+    COMMAND clang-format
+    -style=file
+    -i
+    -verbose
+    ${ALL_SOURCE_FILES}
+  )
+else()
+  message(FATAL_ERROR "clang-format not found. Aborting...")
+endif()
+
+# Adding clang-tidy target if executable is found
+# The configuration below requires cmake 3.6.3+
+# (https://cmake.org/cmake/help/v3.6/variable/CMAKE_LANG_CLANG_TIDY.html)
+find_program(CLANG_TIDY clang-tidy DOC "Path to the clang-tidy tool")
+if(CLANG_TIDY)
+  message("Using clang-tidy")
+  set(CMAKE_CXX_CLANG_TIDY "${CLANG_TIDY};")
+else()
+  message("clang-tidy not found. Proceeding without it...")
+endif()
+
+# Adding "include-what-you-use" target if executable is found
+# The configuration below requires cmake 3.3.2+
+# (https://cmake.org/cmake/help/v3.3/variable/CMAKE_LANG_INCLUDE_WHAT_YOU_USE.html)
+find_program(IWYU include-what-you-use DOC "Path to the include-what-you-use tool")
+if(IWYU)
```

Comment:
Since these tools run as part of the build, rather than via make targets, it could be good to be able to enable them individually, rather than all-or-nothing.  If I'm interested in just the clang-tidy output but happen to have iwyu installed it can be difficult to find the messages of interest.

Reply:
Perhaps `USE_DEV_TOOLS` could do the existence checks and report "<tool> not found" or "<tool> found, add -DUSE_TOOL=ON to enable"?

Reply:
> Perhaps USE_DEV_TOOLS could do the existence checks and report " not found" or " found, add -DUSE_TOOL=ON to enable"?

Would that enable to select which tools are ran? Seems like we'd need to add a flag to turn clang-tidy ON/OFF and same for IWYU, no?

Reply:
Well, there are several ways of doing it, I guess.  We could just have a flag for each tool: `-DUSE_CLANG_TIDY=ON` and `-DUSE_IWYU=ON`.  Or have a top-level flag `-DUSE_DEV_TOOLS=ON` which prints all the info.  The first option is probably better.

I guess the same would apply to the sanitizer.  Could have a single `-DSANITIZER=Address` flag and remove the `USE_SANITIZER`?

Reply:
Ah right, I though you wanted to rename `USE_DEV_TOOLS` to `DUSE_TOOL` which wasn't really enabling to have a finer grained control over which tool gets enabled or not... So what you meant is having options of the form `USE_<TOOL>`, right? If so I just pushed a commit to have this, please have a look

Reply:
Might make sense to put the `if (USE_CLANG_TIDY)` condition at the top level (so no check is done if we don't specify it), and terminate with an error (rather than warning message) if clang-tidy is not found.

Reply:
I agree that would be cleaner indeed. I just pushed a commit for that, thanks @dtebbs